bulletroboy/exoforce/exoforce_hardware.py

```diff
@@ -0,0 +1,266 @@
+			muscle.motor.via_point.world_point = self.end_effectors["right_wrist"]["position"]
+			print("Motor " + str(muscle.id) + " position updated to: " + str(muscle.motor.via_point.world_point))
+
+		print("\nMotor positions updated!\n")
+		self.start_exoforce()
+		self.calibration_timer.destroy()
+
+
+	def print_distance(self):
+		for muscle in self.muscle_units:
+			dist = np.linalg.norm(muscle.motor.via_point.world_point - self.end_effectors["right_wrist"]["position"])
+			if dist < 0.1:
+				self.get_logger().info("Motor " + str(muscle.id) + " touched!")
+			if muscle.id == 3:
+				print(dist)
+
+	def motor_state_listener(self, state):
+		for muscle, displacement in zip(self.muscle_units, state.displacement[:16]):
+			if muscle.displacement is not None:
+				muscle.speed = displacement - muscle.displacement
+			muscle.displacement = displacement
+
+	def collision_listener(self, collision_msg):
+		"""Collision listener.
+
+		Args:
+			collision_msg (Collision): Collision message.
+
+		Returns:
+			-
+		
+		"""
+		self.get_logger().info(f"Received collision: link: {collision_msg.linkid} force: {collision_msg.normalforce}")
+
+		ef = self.decompose_force_link_to_ef(collision_msg.linkid)
+		############ TEMP WORKAROUND
+		# for muscle in self.get_ef_muscle_units(ef):
+		# 	muscle.end_effector.world_point = np.array([0.5, -0.5, -0.3])
+		##########################################################################
+		collision_direction = np.array([collision_msg.contactnormal.x, collision_msg.contactnormal.y, collision_msg.contactnormal.z])
+
+		quaternion = Quaternion(self.end_effectors[ef]["orientation"])
+		force_direction = quaternion.rotation_matrix.dot(collision_direction)
+
+		forces = self.decompose(collision_msg.linkid, collision_msg.normalforce, force_direction)
+		print(forces)
+		
+		for muscle in forces:
+			self.get_muscle_unit(muscle).set_motor_force(forces[muscle])
+
+		motor_ids = []
+		set_points = []
+		for muscle in self.get_ef_muscle_units(ef):
+			motor_ids.append(muscle.id)
+			if forces[muscle.id] == 0:
+				#set_points.append(float(MIN_PWM_TENSION * muscle.direction * -1))
+				set_points.append(0.0)
+			else:
+				set_points.append(self.get_set_point(forces[muscle.id]) * muscle.direction)
+
+		for id, points in zip(motor_ids, set_points):
+			self.get_logger().info("motor " + str(id) + ": " + str(points))
+
+		self.send_motor_commands(motor_ids, set_points)
+		self.applying_force = True
+		self.applying_force_timer.reset()
+
+	def operator_ef_pos_listener(self, ef_pose):
+		"""Callback of the pose subscriber. Sets the pose of the link given in the msg.
+		Args:
+			ef_pose: received PoseStamped msg.
+		
+		Returns:
+			-
+		"""
+		end_effector = ef_pose.header.frame_id
+		################### TEMPORAL MAPPING ############
+		if end_effector == "hand_right":
+			end_effector = "right_wrist"
+		elif end_effector == "hand_left":
+			end_effector = "left_wrist"
+		else:
+			return
+		############################################
```

Comment:
why is it here? dont you have dedicated yamls?

Reply:
see https://github.com/Roboy/bulletroboy/pull/25#issuecomment-760128816